contessa/rules.py

```diff
@@ -60,34 +62,60 @@ def sql_with_where(self):
 
     def apply(self, conn: Connector):
         """
-        Execute a formatted sql. Check if it returns list of booleans that is needed
-        to do a quality check. If yes, return pd.Series.
-        :return: pd.Series
+        Execute a formatted sql. Check if it returns column full of booleans representing validity that is needed
+        to do a quality check. If yes, stream results and return aggregated results
+        :return: AggregatedResult
         """
         sql = self.sql_with_where
         logging.debug(sql)
-        results = [
-            r for r in conn.get_records(sql)
-        ]  # returns generator, so get it to memory
 
-        is_list_of_bool = all(
-            (len(r) == 1 and isinstance(r[0], (bool, type(None))) for r in results)
+        with conn.engine.connect() as con:
+            result = con.execution_options(stream_results=True).execute(sql)
+
+            failed = 0
+            passed = 0
+            failed_examples = set()
+
+            def add_example(row):
+                if len(failed_examples) < failed_example_size:
```

Comment:
shouldnt the user of the method decide whether to add more examples or not?
Also, wouldnt it be neater if the method got the collection of examples as a param, i.e. it would be a method without side consequences, not modifying global state and returning the set instead?

Reply:
- I was intentionally trying not to collect all failed examples in a memory because it could require quite a large chunk for some DQ checks if I did that
- Yes, it would be nicer if user could choose how whether to store an example (e.g. for selection of relevant ones) and how many examples should be kept
- it is not modifying a global state, but a local state of a method wrapping this function
I will look once more at this portion for a ways to improve it.

Reply:
I added an abstract class ExampleSelector and a default implementation FirstNExampleSelector(10) so user can select the examples himself.